lms/djangoapps/experiments/views_custom.py

```diff
@@ -112,33 +115,32 @@ def get(self, request):
                 'upsell_flag': False,
             })
 
-        if 'course_id' not in request.GET:
-            return Response({
-                'show_upsell': False,
-            })
+        course_id = request.GET.get('course_id')
+        try:
+            course_key = CourseKey.from_string(course_id)  # if course_id else None
+        except InvalidKeyError:
+            return HttpResponseBadRequest("Missing or invalid course_id")
 
-        # HACK: the url decoding converts plus to space; put them back
-        course_id = request.GET.get('course_id').replace(' ', '+')
-        course_key = CourseKey.from_string(course_id)
         course = CourseOverview.get_from_id(course_key)
         user = request.user
 
         enrollment = None
-        user_enrollments = None
-        has_non_audit_enrollments = False
+        upsell_mode = True
         try:
-            user_enrollments = CourseEnrollment.objects.select_related('course').filter(user_id=user.id)
-            has_non_audit_enrollments = user_enrollments.exclude(mode__in=CourseMode.UPSELL_TO_VERIFIED_MODES).exists()
             enrollment = CourseEnrollment.objects.select_related(
                 'course'
             ).get(user_id=user.id, course_id=course.id)
+            upsell_mode = enrollment.mode in CourseMode.UPSELL_TO_VERIFIED_MODES
         except CourseEnrollment.DoesNotExist:
             pass  # Not enrolled, use the default values
```

Comment:
nit: I think it would be marginally clearer to set `upsell_mode = False` here.

Reply:
Looking more closely, `upsell_mode` is poorly named; as written, it should be true if the user is not enrolled or if their enrollment mode is among those that can be upsold from.  Thus it defaults to `True` and stays that way if no enrollment for this course is found.

Reply:
I've now renamed this 'user_upsell', added logic to test that the user's enrollment upgrade deadline hasn't passed, and made the flow (hopefully) a little more clear